web/api/tasks.thrift

```diff
@@ -0,0 +1,162 @@
+  ENQUEUED,  // Non-terminated state. Job in the queue, and all inputs are meaningfully available.
+  RUNNING,   // Non-terminated state.
+  COMPLETED, // Terminated state. Successfully ran to completion.
+  FAILED,    // Terminated state. Job was running, but the execution failed.
+  CANCELLED, // Terminated state. Job was cancelled by an administrator, and the cancellation succeeded.
+  DROPPED,   // Terminated state. Job was cancelled due to system reasons (server shutdown, crash, other interference).
+}
+
+struct TaskInfo {
+   1: codechecker_api_shared.TaskToken token,
+   2: string                           taskKind,
+   3: TaskStatus                       status,
+  // If the task is associated with a product, this ID can be used to query
+  // product information, see products.thirft service.
+  // The 'productID' is set to 0 if there is no product associated, meaning
+  // that the task is "global to the server".
+   4: i64                              productId,
+   5: string                           actorUsername,
+   6: string                           summary,
+  // Additional, human-readable comments, history, and log output from the
+  // tasks's processing.
+   7: string                           comments,
+   8: i64                              enqueuedAtEpoch,
+   9: i64                              startedAtEpoch,
+  10: i64                              completedAtEpoch,
+  11: i64                              lastHeartbeatEpoch,
+  // Whether the administrator set this job for a co-operative cancellation.
+  12: bool                             cancelFlagSet,
+}
+
+/**
+ * TaskInfo with additional fields that is sent to administrators only.
+ */
+struct AdministratorTaskInfo {
+  1: TaskInfo normalInfo,
+  2: string   machineId,      // The hopefully unique identifier of the server
+                              // that is/was processing the task.
+  3: bool     statusConsumed, // Whether the main actor of the task
+                              // (see normalInfo.actorUsername) consumed the
+                              // termination status of the job.
+}
+
+/**
+ * Metastructure that holds the filters for getTasks().
+ * The individual fields of the struct are in "AND" relation with each other.
+ * For list<> fields, elements of the list filter the same "column" of the
+ * task information table, and are considered in an "OR" relation.
+ */
+struct TaskFilter {
+   1: list<string>                   tokens,
+   2: list<string>                   machineIDs,
+   3: list<string>                   kinds,
+   4: list<TaskStatus>               statuses,
+  // If empty, it means "all", including those of no username.
+   5: list<string>                   usernames,
+  // If True, it means filter for **only** "no username".
+  // Can not be set together with a non-empty "usernames".
+   6: bool                           filterForNoUsername,
```

Comment:
How about this?
```.thrift
5: optional list<string> usernames
```
In that case `None` and `[]` could be distinguished and `filterForNoUsername` wouldn't be needed. This technique is used for some `ReportFilter` fields, too. I feel it's easier to make a mistake when the client's creator doesn't know that these two fields belong together.
This comment applies for `filterForNoProductID`, too.

Reply:
Alright, I will try reworking the implementation for this...

Reply:
Okay, I implemented the requested changes!